app/vmalert/rule/alerting.go

```diff
@@ -324,18 +330,31 @@ func (ar *AlertingRule) execRange(ctx context.Context, start, end time.Time) ([]
 		prevT := time.Time{}
 		for i := range s.Values {
 			at := time.Unix(s.Timestamps[i], 0)
+			// try to restore alert state from last iteration if needed
+			if at.Equal(start) {
+				if _, ok := ar.alerts[h]; ok {
+					a = ar.alerts[h]
+					prevT = at
+				}
+			}
 			if at.Sub(prevT) > ar.EvalInterval {
 				// reset to Pending if there are gaps > EvalInterval between DPs
 				a.State = notifier.StatePending
 				a.ActiveAt = at
-			} else if at.Sub(a.ActiveAt) >= ar.For {
+				a.Start = time.Time{}
+			} else if at.Sub(a.ActiveAt) >= ar.For && a.State != notifier.StateFiring {
 				a.State = notifier.StateFiring
 				a.Start = at
 			}
 			prevT = at
 			result = append(result, ar.alertToTimeSeries(a, s.Timestamps[i])...)
+			// save alert state if it can be used in next iteration
```

Comment:
Could it be that new alert created earlier wasn't active enough to become active, hence on this moment it will have no state. But in `exec` function (L:445) we'd set it to Pending state. And later will generate ALERTS_FOR_STATE metric for this type. Do we account for this case here?

Reply:
Hmm, not sure I understand the question here. Trying to answer) The `if at.Equal(end)` check here can ensure that if rule has result with end timestamp, it will be saved, no matter its state is `pending` or `firing`.
See https://github.com/VictoriaMetrics/VictoriaMetrics/blob/keep-alert-state-when-replay/app/vmalert/rule/alerting_test.go#L441